### PR TITLE
Use Symbol to represent key names

### PR DESCRIPTION
**Pull request by philwills**

### CODE REVIEW

**Comment by philwills** [comment]
cc @LATaylor-guardian 

I'm not sure whether this is a good idea or not and would value any opinions.  

I quite like the distinction it provides between keys and values.

File: src/main/scala/com/gu/scanamo/ScanamoRequest.scala [hunk 1]
```diff
@@ -41,22 +41,22 @@ object ScanamoRequest {
...
-      .withExpressionAttributeValues(asAVMap(s":$k" -> v))
+      .withKeyConditionExpression(s"#K = :${k.name}")
+      .withExpressionAttributeNames(Map("#K" -> k.name).asJava)
+      .withExpressionAttributeValues(asAVMap(Symbol(s":${k.name}") -> v))
   }
 
-  private def asAVMap[K](kvs: (String, K)*)(implicit fk: DynamoFormat[K]) =
-    Map(kvs: _*).mapValues(fk.write).asJava
+  private def asAVMap[K](kvs: (Symbol, K)*)(implicit fk: DynamoFormat[K]) =
+    Map(kvs: _*).map { case (k,v) => (k.name, fk.write(v)) }.asJava
```
Comment (adamnfish) [inline src/main/scala/com/gu/scanamo/ScanamoRequest.scala]: You use the `->` sugar elsewhere, why not here as well?

File: src/main/scala/com/gu/scanamo/ScanamoRequest.scala [hunk 2]
```diff
@@ -41,22 +41,22 @@ object ScanamoRequest {
...
-  def queryRequest[K](tableName: String)(key: (String, K))(implicit fk: DynamoFormat[K]): QueryRequest = {
+  def queryRequest[K](tableName: String)(key: (Symbol, K))(implicit fk: DynamoFormat[K]): QueryRequest = {
     val (k, v) = key
     new QueryRequest().withTableName(tableName)
-      .withKeyConditionExpression(s"#K = :$k")
-      .withExpressionAttributeNames(Map("#K" -> k).asJava)
-      .withExpressionAttributeValues(asAVMap(s":$k" -> v))
+      .withKeyConditionExpression(s"#K = :${k.name}")
+      .withExpressionAttributeNames(Map("#K" -> k.name).asJava)
+      .withExpressionAttributeValues(asAVMap(Symbol(s":${k.name}") -> v))
```
Comment (adamnfish) [inline src/main/scala/com/gu/scanamo/ScanamoRequest.scala]: Shame there's no `mapKeys` method.

**Comment by adamnfish** [comment]
Much better!

:+1:

File: src/main/scala/com/gu/scanamo/ScanamoRequest.scala (hunk 1 above)
Comment (philwills) [inline src/main/scala/com/gu/scanamo/ScanamoRequest.scala]: When I'm unpacking and repacking a tuple, I like the symmetry of having the bracket form on both sides.  If you could have the `->` sugar in the pattern match, then I'd do that.

**Comment by LATaylor-guardian** [comment]
@philwills I think it makes perfect sense here as a representation for an identifier. Although it does imply some knowledge of what a `Symbol` is so you lose a little clarity somewhat for those not familiar. 

:+1:

File: src/main/scala/com/gu/scanamo/ScanamoRequest.scala (hunk 2 above)
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/scanamo/ScanamoRequest.scala]: I'm assuming implementing filtering on a query request `withFilterExpression` would follow a similar pattern?

Comment (philwills) [inline src/main/scala/com/gu/scanamo/ScanamoRequest.scala]: I'm basically rewriting all of this to support more interesting queries at the moment, but yes, they should hopefully have a similar interface.

**Pull request merged**